Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheV2.kt

```diff
@@ -0,0 +1,944 @@
+     * dispose of this stream
+     * */
+    @SuppressLint("CheckResult")
+    private fun initNormalRxWorkerQueue() {
+        normalRequestQueue
+                .observeOn(workerScheduler)
+                .onBackpressureBuffer()
+                .flatMap { url ->
+                    return@flatMap Flowable.defer { handleFileDownload(url) }
+                        .subscribeOn(workerScheduler)
+                        .onErrorReturn { throwable -> processErrors(url, throwable) }
+                        .map { result -> Pair(url, result) }
+                        .doOnNext { (url, result) -> handleResults(url, result) }
+                }
+                .subscribe({
+                    // Do nothing
+                }, { error ->
+                    throw RuntimeException("Uncaught exception!!! " +
+                            "workerQueue is in error state now!!! " +
+                            "This should not happen!!!, original error = " + error.message)
+                }, {
+                    throw RuntimeException(
+                            "workerQueue stream has completed!!! This should not happen!!!"
+                    )
+                })
+    }
+
+    /**
+     * This is a singleton class so we don't care about the disposable since we will never should
+     * dispose of this stream
+     * */
+    @SuppressLint("CheckResult")
+    private fun initBatchRequestQueue() {
+        batchRequestQueue
+                .observeOn(batchScheduler)
+                .onBackpressureBuffer()
+                .concatMap { urlList ->
+                    return@concatMap Flowable.fromIterable(urlList)
+                        .subscribeOn(batchScheduler)
+                        .concatMap { url ->
+                            return@concatMap handleFileDownload(url)
+                                .onErrorReturn { throwable -> processErrors(url, throwable) }
+                                .map { result -> Pair(url, result) }
+                                .doOnNext { (url, result) ->
+                                    handleResults(url, result)
+                                }
+                        }
+                }
+                .subscribe({
+                    // Do nothing
+                }, { error ->
+                    throw RuntimeException("$TAG Uncaught exception!!! " +
+                            "workerQueue is in error state now!!! " +
+                            "This should not happen!!!, original error = " + error.message)
+                }, {
+                    throw RuntimeException(
+                            "$TAG workerQueue stream has completed!!! This should not happen!!!"
+                    )
+                })
+    }
+
+    fun isRunning(url: String): Boolean {
+        return synchronized(activeDownloads) {
+            activeDownloads.getState(url) == DownloadState.Running
+        }
+    }
+
+    fun enqueueMediaPrefetchRequestBatch(
```

Comment:
This method isn't specific to media prefetching, just call it enqueueRequestBatch.

Reply:
Well, actually, right now it's being used only for prefetching.

Reply:
For album downloads there is enqueueNormalDownloadFileRequest

Reply:
Yeah, the way you did it makes sense with the architecture. Just feels weird that an album download operation isn't a "batch" operation like media prefetching, but that's because it's a batch of ImageSaveTask instances doing it one by one. No changes needed.